server/src/main/java/org/elasticsearch/action/admin/cluster/settings/ClusterGetSettingsAction.java

```diff
@@ -0,0 +1,104 @@
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.action.admin.cluster.settings.RestClusterGetSettingsResponse.PERSISTENT_FIELD;
+import static org.elasticsearch.action.admin.cluster.settings.RestClusterGetSettingsResponse.TRANSIENT_FIELD;
+
+public class ClusterGetSettingsAction extends ActionType<ClusterGetSettingsAction.Response> {
+
+    public static final ClusterGetSettingsAction INSTANCE = new ClusterGetSettingsAction();
+    public static final String NAME = "cluster:admin/settings/get";
+
+    public ClusterGetSettingsAction() {
+        super(NAME, Response::new);
+    }
+
+    /**
+     * Request to retrieve the cluster settings
+     */
+    public static class Request extends MasterNodeReadRequest<Request> {
+        public Request() {}
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+    }
+
+    /**
+     * Response for cluster settings
+     */
+    public static class Response extends ActionResponse implements ToXContentObject {
+        private final Settings persistentSettings;
+        private final Settings transientSettings;
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            persistentSettings = Settings.readSettingsFromStream(in);
+            transientSettings = Settings.readSettingsFromStream(in);
+        }
+
+        public Response(Settings persistentSettings, Settings transientSettings) {
+            this.persistentSettings = Objects.requireNonNullElse(persistentSettings, Settings.EMPTY);
+            this.transientSettings = Objects.requireNonNullElse(transientSettings, Settings.EMPTY);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+
+            builder.startObject(PERSISTENT_FIELD);
+            persistentSettings.toXContent(builder, params);
+            builder.endObject();
+
+            builder.startObject(TRANSIENT_FIELD);
+            transientSettings.toXContent(builder, params);
+            builder.endObject();
+
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            Settings.writeSettingsToStream(persistentSettings, out);
+            Settings.writeSettingsToStream(transientSettings, out);
+        }
+
+        public Settings persistentSettings() {
+            return persistentSettings;
+        }
+
+        public Settings transientSettings() {
+            return transientSettings;
+        }
+
+        public Settings settings() {
+            return Settings.builder().put(persistentSettings).put(transientSettings).build();
```

Comment:
I'm not sure this is the right way to get the combined settings, it didn't make sense to double serialize across the transport protocol.

Reply:
It looks right:

https://github.com/elastic/elasticsearch/blob/579156717a284091dfe3810769a2a1194f6b76e8/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L1875

However I think it would be fine (preferable even) to send the combined settings separately from the transient and persistent ones, rather than duplicating this logic and risking getting it wrong.